### DISCUSSION TITLE
could we use username and password to create a webhook by rest api

### DISCUSSION THREAD

**Discussion by Amykkkkkk** [community/community#39137] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Here is the example of creating a repository webhook.
curl \
  -X POST \
  -H "Accept: application/vnd.github+json" \
  **-H "Authorization: Bearer <YOUR-TOKEN>**" \
  http(s)://HOSTNAME/api/v3/repos/OWNER/REPO/hooks \
  -d '{"name":"web","active":true,"events":["push","pull_request"],"config":{"url":"https://example.com/webhook","content_type":"json","insecure_ssl":"0"}}'

Is this token an access token?
Is it possible to use username and password instead of token?

**Comment by airtower-luna** [comment, marked as answer]
> Is this token an access token?

Yes, there has to be a token with appropriate permissions in the `Authorization` header (after `Bearer`).

> Is it possible to use username and password instead of token?

No.

**Reply by Amykkkkkk** [reply]
Thanks a lot in advance.
 
I'm new to this so I also have some other questions about this api:

1. it's just a post request.Does it mean I just need to set once and then it works?
2. about the parameter: http(s)://HOSTNAME/api/v3/repos/OWNER/REPO/hooks, if I have a  organization aaa and there is a repo bbb in this organization and the github api url is https://github.xxx/api/v3 and owner is user1, so the correct parameter is https://github.xxx/api/v3/repos/use1/aaa/bbb/hooks?
3. I could use this curl command anywhere only if I know the token and url, right?
4. besides, I need to use a secret during creating the webhook. could I use the secret value which is in the secret.yaml in k8s?

Suddenly I felt defeated...

**Reply by airtower-luna** [reply]
> it's just a post request.Does it mean I just need to set once and then it works?

It should, the same way as if you configure a webhook over the web UI. :slightly_smiling_face:

> about the parameter: http(s)://HOSTNAME/api/v3/repos/OWNER/REPO/hooks, if I have a organization aaa and there is a repo bbb in this organization and the github api url is https://github.xxx/api/v3 and owner is user1, so the correct parameter is https://github.xxx/api/v3/repos/use1/aaa/bbb/hooks?

No, the owner of the repository is the organization. The user(s) who own the organization don't show up in the URL. Replace `{owner}` in the URL pattern with the organization name.

> I could use this curl command anywhere only if I know the token and url, right?

Exactly. The token is what allows you to make the change.

> besides, I need to use a secret during creating the webhook. could I use the secret value which is in the secret.yaml in k8s?

I don't know your k8s infrastructure, so I can't say for sure, but probably not. The secret (if configured) is used for [integrity protection of the webhook data](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#delivery-headers), so if you use it you should use a dedicated value, not one you already use for other purposes. But obviously that value should be part of your config, because you'll need it to verify the data.

**Reply by Amykkkkkk** [reply]
Thank you very much!!!!!

I just realized where did I go wrong several minutes ago. 
In the past several hours, I tried some other github rest api by postman and they succeeded. I didn't know that **the owner of the repository is the organization** until observing those responses. 

About secret part, I haven't figured out their relationships between the parameter in this api with the secret.yaml in k8s so I plan to do the test after nine hours. hhh.. Although I'm excited now, I have to sleep or I won't be able to get up tomorrow.

You are so nice. Your explanation gave me much support and help, which inspires me to do the practise by postman.
Thanks again for your nice explanation and suggestions!!!
Have a nice day~

**Reply by airtower-luna** [reply]
Thank you very much for your sweet words, and I hope you got it working! :smile_cat:
